perf/src/server.rs

```diff
@@ -0,0 +1,160 @@
+                quinn::CertificateChain::from_pem(&cert).context("parsing cert")?,
+            )
+        }
+        _ => {
+            let cert = rcgen::generate_simple_self_signed(vec!["localhost".into()]).unwrap();
+            (
+                quinn::PrivateKey::from_der(&cert.serialize_private_key_der()).unwrap(),
+                quinn::CertificateChain::from_certs(vec![quinn::Certificate::from_der(
+                    &cert.serialize_der().unwrap(),
+                )
+                .unwrap()]),
+            )
+        }
+    };
+
+    let mut server_config = quinn::ServerConfigBuilder::default();
+    server_config.certificate(cert, key).unwrap();
+    server_config.protocols(&[b"perf"]);
+
+    let server_config = server_config.build();
+
+    let mut endpoint = quinn::EndpointBuilder::default();
+    endpoint.listen(server_config);
+
+    let (endpoint, mut incoming) = endpoint.bind(&opt.listen).context("binding endpoint")?;
+
+    info!("listening on {}", endpoint.local_addr().unwrap());
+
+    while let Some(handshake) = incoming.next().await {
+        tokio::spawn(async move {
+            if let Err(e) = handle(handshake).await {
+                error!("connection lost: {:#}", e);
+            }
+        });
+    }
+
+    Ok(())
+}
+
+async fn handle(handshake: quinn::Connecting) -> Result<()> {
+    let quinn::NewConnection {
+        uni_streams,
+        bi_streams,
+        connection,
+        ..
+    } = handshake.await.context("handshake failed")?;
+    debug!("{} connected", connection.remote_address());
+    tokio::select! {
+        x = drive_uni(connection, uni_streams) => x?,
+        x = drive_bi(bi_streams) => x?,
+    }
+    Ok(())
+}
+
+async fn drive_uni(
+    connection: quinn::Connection,
+    mut streams: quinn::IncomingUniStreams,
+) -> Result<()> {
+    while let Some(stream) = streams.next().await {
+        let stream = stream?;
+        let connection = connection.clone();
+        tokio::spawn(async move {
+            if let Err(e) = handle_uni(connection, stream).await {
+                error!("request failed: {:#}", e);
+            }
+        });
+    }
+    Ok(())
+}
+
+async fn handle_uni(connection: quinn::Connection, stream: quinn::RecvStream) -> Result<()> {
+    let bytes = read_req(stream).await?;
+    let response = connection.open_uni().await?;
+    respond(bytes, response).await?;
+    Ok(())
+}
+
+async fn drive_bi(mut streams: quinn::IncomingBiStreams) -> Result<()> {
+    while let Some(stream) = streams.next().await {
+        let (send, recv) = stream?;
+        tokio::spawn(async move {
+            if let Err(e) = handle_bi(send, recv).await {
+                error!("request failed: {:#}", e);
+            }
+        });
+    }
+    Ok(())
+}
+
+async fn handle_bi(send: quinn::SendStream, recv: quinn::RecvStream) -> Result<()> {
+    let bytes = read_req(recv).await?;
+    respond(bytes, send).await?;
```

Comment:
draining of the request stream is still missing.  What I'm not sure is whether this actually should happen concurrently to sending a request. I implemented it now somewhere else as:

```
let bytes = read_header(&mut recv).await?;

let drain_request_fut = drain_data(recv);
let send_response_fut = send_response(bytes, send);

let ((), ()) = tokio::try_join!(drain_request_fut, send_response_fut)?;

Ok(())
```

Reply:
Reading through the update https://github.com/nibanks/quic-perf/pull/2/files , concurrent sending of the response doesn't seem required and the `try_join!` can be dropped.